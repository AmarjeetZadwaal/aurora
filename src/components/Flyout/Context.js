import React from "react";

const { Provider, Consumer } = React.createContext();

export { Provider as FlyoutProvider };
export { Consumer as FlyoutConsumer };
