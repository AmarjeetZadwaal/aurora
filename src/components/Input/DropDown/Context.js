import React from "react";

export const { Provider, Consumer } = React.createContext({
  isOpen: null,
  onClose: null
});
