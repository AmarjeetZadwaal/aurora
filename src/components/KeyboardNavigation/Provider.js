import React from "react";
import PropTypes from "prop-types";
import { Provider } from "./Context";

const ARROWUP = 38;
const ARROWDOWN = 40;
const SPACEBAR = 32;
const TABKEY = 9;
const ENTERKEY = 13;

const WHITELIST = [ARROWUP, ARROWDOWN, SPACEBAR, TABKEY, ENTERKEY];

export default class KeyBoardProvider extends React.Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    role: PropTypes.string.isRequired
  };
  onClick = ({ index }) => {
    this.setState(() => ({ focused: index }));
  };

  onKeyDown = event => {
    if (!WHITELIST.includes(event.keyCode)) return;

    /* eslint-disable */
    ![TABKEY, ENTERKEY, SPACEBAR].includes(event.keyCode) &&
      event.preventDefault();
    /* eslint-enable */

    this.cycleSelect(event.keyCode);
  };

  cycleSelect = key => {
    const { focused } = this.state;
    const { children } = this.props;

    if (key === ARROWUP && focused <= 0) {
      this.setState(() => ({ focused: children.length - 1 }));
    }

    if (key === ARROWUP && focused > 0) {
      this.setState(prevState => ({ focused: prevState.focused - 1 }));
    }

    if (key === ARROWDOWN && focused >= children.length - 1) {
      this.setState(() => ({ focused: 0 }));
    }

    if (key === ARROWDOWN && focused < children.length - 1) {
      this.setState(prevState => ({ focused: prevState.focused + 1 }));
    }
  };

  /* eslint-disable */
  state = {
    focused: null,
    focusSelected: this.onClick
  };
  /* eslint-enable */

  render() {
    const { role } = this.props;
    return (
      <Provider value={this.state}>
        {/* eslint-disable */}
        <div role={role} onKeyDown={this.onKeyDown}>
          {this.props.children}
        </div>
        {/* eslint-enable */}
      </Provider>
    );
  }
}
