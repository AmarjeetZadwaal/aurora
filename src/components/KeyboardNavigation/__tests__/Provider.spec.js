import React from "react";
import { renderIntoDocument, cleanup, fireEvent } from "react-testing-library";

import Provider from "../Provider";
import { Consumer } from "../Context";

const TesingConsumers = (
  { children, index, datatestid } // eslint-disable-line
) => (
  <Consumer>
    {({ focused }) => (
      <div data-isfocues={focused === index} data-testid={datatestid}>
        {children}
      </div>
    )}
  </Consumer>
);

describe("KeyBoardProvider", () => {
  afterEach(cleanup);

  it("Up outside range of number of Children ", () => {
    const { getByTestId, container } = renderIntoDocument(
      <Provider role="option">
        <TesingConsumers datatestid="button0" index={0}>
          Testing1
        </TesingConsumers>
        <TesingConsumers datatestid="button1" index={1}>
          Testing2
        </TesingConsumers>
        <TesingConsumers datatestid="button2" index={2}>
          Testing3
        </TesingConsumers>
        <TesingConsumers datatestid="button3" index={3}>
          Testing4
        </TesingConsumers>
      </Provider>
    );

    fireEvent.keyDown(getByTestId("button0"), { keyCode: 38 });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("Up within range of number of Children", () => {
    const { getByTestId, container } = renderIntoDocument(
      <Provider role="option">
        <TesingConsumers datatestid="button0" index={0}>
          Testing1
        </TesingConsumers>
        <TesingConsumers datatestid="button1" index={1}>
          Testing2
        </TesingConsumers>
        <TesingConsumers datatestid="button2" index={2}>
          Testing3
        </TesingConsumers>
        <TesingConsumers datatestid="button3" index={3}>
          Testing4
        </TesingConsumers>
      </Provider>
    );

    fireEvent.keyDown(getByTestId("button0"), { keyCode: 38 });
    fireEvent.keyDown(getByTestId("button0"), { keyCode: 38 });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("Down within range of number of Children", () => {
    const { getByTestId, container } = renderIntoDocument(
      <Provider role="option">
        <TesingConsumers datatestid="button0" index={0}>
          Testing1
        </TesingConsumers>
        <TesingConsumers datatestid="button1" index={1}>
          Testing2
        </TesingConsumers>
        <TesingConsumers datatestid="button2" index={2}>
          Testing3
        </TesingConsumers>
        <TesingConsumers datatestid="button3" index={3}>
          Testing4
        </TesingConsumers>
      </Provider>
    );

    fireEvent.keyDown(getByTestId("button0"), { keyCode: 40 });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("Down outside range of number of Children ", () => {
    const { getByTestId, container } = renderIntoDocument(
      <Provider role="option">
        <TesingConsumers datatestid="button0" index={0}>
          Testing1
        </TesingConsumers>
        <TesingConsumers datatestid="button1" index={1}>
          Testing2
        </TesingConsumers>
        <TesingConsumers datatestid="button2" index={2}>
          Testing3
        </TesingConsumers>
        <TesingConsumers datatestid="button3" index={3}>
          Testing4
        </TesingConsumers>
      </Provider>
    );

    fireEvent.keyDown(getByTestId("button0"), { keyCode: 40 });
    fireEvent.keyDown(getByTestId("button0"), { keyCode: 40 });
    fireEvent.keyDown(getByTestId("button0"), { keyCode: 40 });
    fireEvent.keyDown(getByTestId("button0"), { keyCode: 40 });
    fireEvent.keyDown(getByTestId("button0"), { keyCode: 40 });

    expect(container.firstChild).toMatchSnapshot();
  });
  it("Down outside range of number of Children ", () => {
    const { getByTestId, container } = renderIntoDocument(
      <Provider role="option">
        <TesingConsumers datatestid="button0" index={0}>
          Testing1
        </TesingConsumers>
        <TesingConsumers datatestid="button1" index={1}>
          Testing2
        </TesingConsumers>
        <TesingConsumers datatestid="button2" index={2}>
          Testing3
        </TesingConsumers>
        <TesingConsumers datatestid="button3" index={3}>
          Testing4
        </TesingConsumers>
      </Provider>
    );

    fireEvent.keyDown(getByTestId("button0"), { keyCode: 3 });

    expect(container.firstChild).toMatchSnapshot();
  });
});
