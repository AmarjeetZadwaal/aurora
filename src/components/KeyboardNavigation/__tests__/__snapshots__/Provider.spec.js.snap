// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyBoardProvider Down outside range of number of Children  1`] = `
<div
  role="option"
>
  <div
    data-isfocues="false"
    data-testid="button0"
  >
    Testing1
  </div>
  <div
    data-isfocues="true"
    data-testid="button1"
  >
    Testing2
  </div>
  <div
    data-isfocues="false"
    data-testid="button2"
  >
    Testing3
  </div>
  <div
    data-isfocues="false"
    data-testid="button3"
  >
    Testing4
  </div>
</div>
`;

exports[`KeyBoardProvider Down outside range of number of Children  2`] = `
<div
  role="option"
>
  <div
    data-isfocues="false"
    data-testid="button0"
  >
    Testing1
  </div>
  <div
    data-isfocues="false"
    data-testid="button1"
  >
    Testing2
  </div>
  <div
    data-isfocues="false"
    data-testid="button2"
  >
    Testing3
  </div>
  <div
    data-isfocues="false"
    data-testid="button3"
  >
    Testing4
  </div>
</div>
`;

exports[`KeyBoardProvider Down within range of number of Children 1`] = `
<div
  role="option"
>
  <div
    data-isfocues="false"
    data-testid="button0"
  >
    Testing1
  </div>
  <div
    data-isfocues="true"
    data-testid="button1"
  >
    Testing2
  </div>
  <div
    data-isfocues="false"
    data-testid="button2"
  >
    Testing3
  </div>
  <div
    data-isfocues="false"
    data-testid="button3"
  >
    Testing4
  </div>
</div>
`;

exports[`KeyBoardProvider Up outside range of number of Children  1`] = `
<div
  role="option"
>
  <div
    data-isfocues="false"
    data-testid="button0"
  >
    Testing1
  </div>
  <div
    data-isfocues="false"
    data-testid="button1"
  >
    Testing2
  </div>
  <div
    data-isfocues="false"
    data-testid="button2"
  >
    Testing3
  </div>
  <div
    data-isfocues="true"
    data-testid="button3"
  >
    Testing4
  </div>
</div>
`;

exports[`KeyBoardProvider Up within range of number of Children 1`] = `
<div
  role="option"
>
  <div
    data-isfocues="false"
    data-testid="button0"
  >
    Testing1
  </div>
  <div
    data-isfocues="false"
    data-testid="button1"
  >
    Testing2
  </div>
  <div
    data-isfocues="true"
    data-testid="button2"
  >
    Testing3
  </div>
  <div
    data-isfocues="false"
    data-testid="button3"
  >
    Testing4
  </div>
</div>
`;
